pkg/agent/qrm-plugins/memory/dynamicpolicy/util_linux_amd64.go

```diff
@@ -93,3 +117,371 @@ containerLoop:
+
+		if len(leftPhyPages) > movePagesEachTime {
+			movingPagesAddr = leftPhyPages[:movePagesEachTime]
+			leftPhyPages = leftPhyPages[movePagesEachTime:]
+		} else {
+			movingPagesAddr = leftPhyPages[:]
+			leftPhyPages = leftPhyPages[:0]
+		}
+
+		nodes := make([]int32, len(movingPagesAddr))
+		for i := range movingPagesAddr {
+			nodes[i] = int32(dstNuma)
+		}
+
+		start := time.Now()
+		_, err := moveProcessPages(pid, uint64(len(movingPagesAddr)), movingPagesAddr, nodes)
+		if err != nil {
+			errList = append(errList, fmt.Errorf("failed move pages for pid: %d, numa: %d err: %v", pid, dstNuma, err))
+			continue
+		}
+		timeCost := time.Since(start).Milliseconds()
+
+		if timeCost == 0 {
+			movePagesEachTime = MovePagesMaxEachTime
+			continue
+		}
+
+		movePagesEachTime = movePagesEachTime * MovePagesAcceptableTimeCost / int(timeCost)
+		if movePagesEachTime < MovePagesMinEachTime {
+			movePagesEachTime = MovePagesMinEachTime
+		} else if movePagesEachTime > MovePagesMaxEachTime {
+			movePagesEachTime = MovePagesMaxEachTime
+		}
+	}
+
+	return utilerrors.NewAggregate(errList)
+}
+
+func getNumaNodeFreeMemMB(nodeID int) (uint64, error) {
+	nodeVmstatFile := fmt.Sprintf("/sys/devices/system/node/node%d/vmstat", nodeID)
+	lines, err := general.ReadFileIntoLines(nodeVmstatFile)
+	if err != nil {
+		e := fmt.Errorf("failed to ReadFile %s, err %s", nodeVmstatFile, err)
+		return 0, e
+	}
+
+	for _, line := range lines {
+		if line == "" {
+			continue
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) == 0 {
+			continue
+		}
+
+		if len(fields) != 2 {
+			return 0, fmt.Errorf("invalid line %s in vmstat file %s", line, nodeVmstatFile)
+		}
+
+		if fields[0] != "nr_free_pages" {
+			continue
+		}
+
+		val, err := strconv.ParseUint(fields[1], 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("invalid line %s in vmstat file %s, err %s", line, nodeVmstatFile, err)
+		}
+
+		return val * 4 / 1024, nil
+	}
+
+	return 0, fmt.Errorf("failed to find nr_free_pages in %s", nodeVmstatFile)
+}
+
+// ratio range [1, 10], so return value is an approximate value
+func getNumasFreeMemRatio(numas []int) (map[int]int, error) {
+	var totalSize uint64
+	numaFreeMem := make(map[int]uint64)
+	for _, n := range numas {
+		freeMemSize, err := getNumaNodeFreeMemMB(n)
+		if err != nil {
+			return nil, fmt.Errorf("failed to getNumaNodeFreeMem for node %d, err %s", n, err)
+		}
+
+		numaFreeMem[n] = freeMemSize
+		totalSize += freeMemSize
+	}
+
+	// convert numa free memory ratio to a approximate value which is in the range [1, 10]
+	var fraction uint64 = 1
+	if totalSize > 10 {
+		fraction = (totalSize + 10) / 10
+	}
+
+	numaFreeMemRatio := make(map[int]int)
+	for numaID, freeMemSize := range numaFreeMem {
+		val := freeMemSize / fraction
+		if val == 0 {
+			continue
+		}
+		numaFreeMemRatio[numaID] = int(val)
+	}
+
+	return numaFreeMemRatio, nil
+}
+
+func getProcessPageStats(pid int) (*smapsInfo, error) {
+	smapFile := fmt.Sprintf("/proc/%d/smaps", pid)
```

Comment:
should we allow this action? should we use malachite instead of read metrics here?

Reply:
leave to  further discussion in future for metrics between katalyst and malachite